### ISSUE TITLE
Static analysis issue regarding ReadableCollection interface

### ISSUE THREAD

**Issue by DesLynx** [doctrine/collections#337]
Hi! 
After upgrading from 1.7.3 to 1.8.0 PHPStan analysis (max level) returned a flood of error message from (previously) perfectly valid code because it seems that when this type of declaration was enough for the generic part of the `Collection` interface: 
```php
    /**
     * @var Collection<int, MyObjectType>
     */
    private Collection $myObjectCollection;
```
it now needs a declaration for ReadableCollection interface:
```php
    /**
     * @var Collection<int, MyObjectType>|ReadableCollection<int, MyObjectType>
     */
    private Collection|ReadableCollection $myObjectCollection;
```

else a call of method from `ReadableCollection` interface like say `getValues()` is not analyzed properly as returning an array of `MyObjectType` but as an array of `mixed`. More generally it seems that the declaration of `TKey, T` for the `Collection` interface is not known by `ReadableCollection` interface

I'm not an expert in static analysis so I'm wondering if this is normal or if there is an issue with interface heritage regarding PHPStan or psalm.

**Comment by greg0ire** [comment]
Are you getting similar issues with Psalm? Are you using an up-to-date version of PHPStan?
I believe this should be more than enough to do the trick: https://github.com/doctrine/collections/blob/ca783dc3ff8afe806a16f118b9e4d42cb71b28aa/lib/Doctrine/Common/Collections/Collection.php#L27

**Comment by DesLynx** [comment]
phpstan/phpstan -> 1.8.6 (last version)
phpstan/phpstan-doctrine -> 1.3.14 (last version)
Not tested with Psalm yet

**Comment by greg0ire** [comment]
> a flood of error message

It would be helpful to provide at least one of them.

**Comment by DesLynx** [comment]
Psalm level 1 doesn't report new errors in my code between version 1.7.3 and 1.8.0 of the package. 
Code sample + phpstan error messages to follow...

**Comment by DesLynx** [comment]
The following code:
```php
$userParticipating = $event->getEventParticipants()
    ->filter(fn (EventParticipant $ev) => EventParticipant::PARTICIPATING === $ev->getState())
    ->map(fn (EventParticipant $ev) => $ev->getUser())
    ->getValues();
```

Produce the following errors:
```
- Parameter #1 $p of method Doctrine\Common\Collections\Collection<int,App\Entity\EventParticipant>::filter() expects Closure(mixed): bool, Closure(App\Entity\EventParticipant): bool given.                             
- Parameter #1 $func of method Doctrine\Common\Collections\ReadableCollection<(int|string),mixed>::map() expects Closure(mixed): App\Entity\User|null, Closure(App\Entity\EventParticipant): App\Entity\User|null given.
``` 

Extract of the related class:
```php

    /**
     * @var Collection<int, EventParticipant>|EventParticipant[]
     */
    #[ORM\OneToMany(mappedBy: 'event', targetEntity: EventParticipant::class, cascade: ['persist'], orphanRemoval: true)]
    private Collection|iterable $eventParticipants;
...
    /**
     * @return Collection<int, EventParticipant>|EventParticipant[]
     */
    public function getEventParticipants(): Collection|iterable
    {
        return $this->eventParticipants;
    }

    public function addEventParticipant(EventParticipant $eventParticipant): self
    {
        if (!$this->eventParticipants->contains($eventParticipant)) {
            $this->eventParticipants[] = $eventParticipant;
            $eventParticipant->setEvent($this);
        }

        return $this;
    }

    public function removeEventParticipant(EventParticipant $eventParticipant): self
    {
        // set the owning side to null (unless already changed)
        if ($this->eventParticipants->removeElement($eventParticipant) && $eventParticipant->getEvent() === $this) {
            $eventParticipant->setEvent(null);
        }

        return $this;
    }

```

**Comment by greg0ire** [comment]
Do you still observe that particular error when you remove `phpstan-doctrine`? I'm wondering if this file might not break everything: https://github.com/phpstan/phpstan-doctrine/blob/1.3.x/stubs/Collections/Collection.stub

**Comment by DesLynx** [comment]
Good guess! Removing the phpstan-doctrine removed the errors.

**Comment by DesLynx** [comment]
Feel free to close this issue I will open one at phpstan-doctrine.

**Comment by DesLynx** [comment]
And many thanks for the time you take to investigate.

**Comment by greg0ire** [comment]
Phew! You had me worried for a moment there, I must admit! The fact that you were the only one reporting this was definitely a hint :)

**Comment by ondrejmirtes** [comment]
Fixed in phpstan-doctrine 1.3.15. Please test it and let me know :)

**Comment by DesLynx** [comment]
The fix in phpstan-doctrine resolved my issue
